### DISCUSSION TITLE
Where do I begin to understand how xonsh terminal/pseudoterminal/PTY works?

### DISCUSSION THREAD

**Discussion by daddycocoaman** [xonsh/xonsh#4216] [General]
This is more of a request to help understand some code and get some ideas rather than any standing issue.

I've spent the last week or so looking around the codebase to understand how xonsh works as a shell but I'm having a difficult time getting the hang of it. I am trying to work through an idea of being able to have a **_custom_** remote Xonsh server that I can connect to and run commands against (implying I would fork and customize for myself if I have to).

I know xxh sort of touches on this but I am actually looking to perform this against Windows hosts, which xxh does not support. Ideally, I would connect to some server port where a PTY for that remote Xonsh would launch and I could perform normal actions.  I've had a **little** success using sockets and [pywinpty](https://github.com/spyder-ide/pywinpty) but I was using netcat to connect to the port. It would make more sense for me to use a Xonsh client to connect to a Xonsh server!

Unfortunately, I have not yet figured out how to properly use a pty with xonsh as a remote server with full functionality. Has anyone seen any concept like this before? Any suggestions where I could dig even further to see how using a xonsh client/server configuration could possibly work? Thanks for any help or ideas.

**Comment by anki-code** [comment]
The xonsh shell is using [prompt-toolkit](https://github.com/prompt-toolkit/python-prompt-toolkit) library.

**Reply by daddycocoaman** [reply]
Right, but xonsh does a lot with how it handles input and output in a way that's different than a lot of other PTK based tools, particularly when it comes to things like in-line functions. 

So with my testing with netcat, for example, pressing Enter would send the input and xonsh would process that line as the entire input, even if I meant to just move to the next line to write a function.

The next idea I have is trying to wrapping the client/socket connection with Paramiko for a better TTY experience but I was wondering if anyone had just seen this client/server concept before so I could learn from previous experience, or if there were any parts of xonsh I should look around.

If not, no worries....I'll just close this and report back later if I can do this successfully.

**Reply by anki-code** [reply]
You can start from `cmdloop()` to understand how prompt-toolkit prompt works in xonsh:

https://github.com/xonsh/xonsh/blob/5a792a6eec4889cbbc50619da6b26f7471d7ce5d/xonsh/ptk_shell/shell.py#L330-L331

Client/server. I remember that @scopatz tried to do something similar when he tried to create the live xonsh demo but he is vacationing now. May be @gforsyth can remember.
